com.unity.ml-agents/Runtime/Sensors/ObservationSpec.cs

```diff
@@ -0,0 +1,108 @@
+using Unity.Barracuda;
+
+namespace Unity.MLAgents.Sensors
+{
+    /// <summary>
+    /// A description of the observations that an ISensor produces.
+    /// This includes the size of the observation, the properties of each dimension, and how the observation
+    /// should be used for training.
+    /// </summary>
+    public struct ObservationSpec
+    {
+        /// <summary>
+        /// The size of the observations that will be generated.
+        /// For example, a sensor that observes the velocity of a rigid body (in 3D) would use [3].
+        /// A sensor that returns an RGB image would use [Height, Width, 3].
+        /// </summary>
+        public InplaceArray<int> Shape;
```

Comment:
Open to discussion on the visibility for these fields, and also the constructor. I think the static helpers (Vector, VariableLength, Visual) are good for all the purposes that we currently support in the trainer, and a user could potentially implement their own trainer that supported different combos of properties (e.g. convolutional 1D or 3D).

Reply:
I think we want to expose the ObservationSpec() constructor, and with that it's flexible enough for users to customize their own and not necessary to expose the fields

We check in the constructor that the length of shape and dimensionProperties match. Users can potentially set shape and dimensionProperties with different length if we expose the fields.

Reply:
OK, you convinced me. I'll make the fields internal and add public get properties.

Reply:
Updated.